distsys/resources/tcpmailboxes.go

```diff
@@ -0,0 +1,350 @@
+							// at least, in this case the msgChannel will function as a rate limiter, so
+							// crash-free operation shouldn't do anything weird
+
+							// a restart-proof method would take advantage of TCP necessarily dropping the connection,
+							// thus ending this connection, and log enough that everything important can be recovered
+							err = encoder.Encode(false)
+							if err != nil {
+								continue
+							}
+							for _, elem := range localBuffer {
+								msgChannel <- elem
+							}
+							localBuffer = nil
+							hasBegun = false
+						}
+					}
+					err = conn.Close()
+					if err != nil {
+						fmt.Printf("error closing connection: %v", err)
+					}
+				}()
+			}
+		}()
+
+		return &tcpMailboxesLocalArchetypeResource{
+			msgChannel: msgChannel,
+		}
+	})
+}
+
+func (res *tcpMailboxesLocalArchetypeResource) Abort() chan struct{} {
+	res.readBacklog = append(res.readsInProgress, res.readBacklog...)
+	res.readsInProgress = nil
+	return nil
+}
+
+func (res *tcpMailboxesLocalArchetypeResource) PreCommit() chan error {
+	return nil
+}
+
+func (res *tcpMailboxesLocalArchetypeResource) Commit() chan struct{} {
+	res.readsInProgress = nil
+	return nil
+}
+
+func (res *tcpMailboxesLocalArchetypeResource) ReadValue() (distsys.TLAValue, error) {
+	// if a critical section previously aborted, already-read values will be here
+	if len(res.readBacklog) > 0 {
+		value := res.readBacklog[0]
+		res.readBacklog[0] = distsys.TLAValue{} // ensure this TLAValue is null, otherwise it will dangle and prevent potential GC
+		res.readBacklog = res.readBacklog[1:]
+		res.readsInProgress = append(res.readsInProgress, value)
+		return value, nil
+	}
+
+	// otherwise, either pull a notification + atomically read a value from the buffer, or time out
+	select {
+	case msg := <-res.msgChannel:
+		res.readsInProgress = append(res.readsInProgress, msg)
+		return msg, nil
+	case <-time.After(tcpMailboxesSendTimeout):
+		return distsys.TLAValue{}, distsys.ErrCriticalSectionAborted
+	}
+}
+
+func (res *tcpMailboxesLocalArchetypeResource) WriteValue(value distsys.TLAValue) error {
+	panic(fmt.Errorf("attempted to write value %v to a local mailbox archetype resource", value))
+}
+
+type tcpMailboxesRemoteArchetypeResource struct {
+	distsys.ArchetypeResourceLeafMixin
+	dialAddr string
+
+	inCriticalSection bool
+	conn              net.Conn
+	connEncoder       *gob.Encoder
+	connDecoder       *gob.Decoder
+}
+
+var _ distsys.ArchetypeResource = &tcpMailboxesRemoteArchetypeResource{}
+
+func tcpMailboxesRemoteArchetypeResourceMaker(dialAddr string) distsys.ArchetypeResourceMaker {
+	return distsys.ArchetypeResourceMakerFn(func() distsys.ArchetypeResource {
+		return &tcpMailboxesRemoteArchetypeResource{
+			dialAddr: dialAddr,
+		}
+	})
+}
+
+func (res *tcpMailboxesRemoteArchetypeResource) ensureConnection() error {
+	if res.conn == nil {
+		var err error
+		res.conn, err = net.Dial("tcp", res.dialAddr)
```

Comment:
Note that we might get blocked here indefinitely.

Reply:
This surprises me. Can you describe the scenario?

Reply:
(this is different to below, because I don't even know about a timeout mechanism for Dial)

Reply:
From `net.Dial` documentation, I didn't understand that it timeouts. However, `net` has `DialTimeout` and `DialContext` function that explicitly have timeouts.